src/bytecode/mod.rs

```diff
@@ -32,10 +42,44 @@ impl<'a> LuaBytecode {
     /// Generates a fresh register and returns it.
     /// This is used in cases like `x = 1 + 2 + 3` to generate intermmediate
     /// registers in which, for instance, the result of 2 + 3 is stored.
-    pub fn get_new_reg(&mut self) -> String {
+    pub fn get_new_reg(&mut self) -> u64 {
         let reg = self.reg_counter;
+        self.registers.push(Reg::new(reg));
         self.reg_counter += 1;
-        format!("${}", reg).to_string()
+        self.registers.last().unwrap().id()
+    }
+
+    /// Get the register that corresponds to the given identifier.
+    /// If the corresponding register is not found, a new register is created
+    /// and returned.
+    pub fn get_reg(&mut self, id: &str) -> u64 {
+        if let Some(id) = self.reg_map.get(id) {
+            return *id;
+        }
+        let reg_id = self.get_new_reg();
+        self.reg_map.insert(id.to_string().clone(), reg_id);
+        reg_id
+
+    }
+
+    /// Set the register to the given value
+    pub fn set_value(&mut self, id: u64, value: Value) {
+        let index = self.find_reg(id);
+        self.registers[index].set_value(value);
+    }
+
+    pub fn get_value(&self, id: u64) -> &Value {
+        let index = self.find_reg(id);
+        self.registers[index].get_value()
+    }
+
+    fn find_reg(&self, id: u64) -> usize {
+        let dummy_reg = Reg::new(id);
+        if let Ok(index) = self.registers.as_slice().binary_search(&dummy_reg) {
```

Comment:
Why do we need a binary search?

Reply:
Because I am trying to find the given id in the vector of registers. I know that the registers are always sorted, so binary search sounds reasonable. I could have opted for a `HashMap<u64, Reg>` instead, but this approach is going to use more space (because the Register already keeps track of an ID). Do you think this approach might be better?

Reply:
Binary search is fast... but not *that* fast. This routine, if it's called regularly, is going to be a horrible bottleneck. Again, maybe this is something for a future PR. But, in general, we want to do everything with integer indices into a vector (can't get quicker than that!).

Reply:
Ah I see, I will leave this for the next PR then.

Reply:
I was being stupid, the binary search is wrong. I can just index the vector directly...